In [111]:
import numpy as np
import matplotlib.pyplot as plt

- Section 1 : Linear Regression
    - We will use the Boston_housing dataset for the regression problem. Run the below cell to get the following variables:

        - Training_data = Training data matrix of shape  (n,d) 
        - labels = label vector corresponding to the training data
        - test_data = Test data matrix of shape  (n1,d)  where  n1  is the number of examples in test dataset.
        - test_labels = label vector corresponding to the test data
        - Use this dataset for the regression problem.

In [112]:
from tensorflow.keras.datasets import boston_housing
Train, test = boston_housing.load_data(seed= 111)
Training_data, labels = Train[0], Train[1]
Test_data, test_labels = test[0], test[1]

- Train , Test: ye es tuple hota hai... jisme do element hote hai (Data_Matrix , label_array) 

- then Jab hamne Training_data , labels = Train[0] , Train[1] kiya to Training_data me Data_Matrix chala gaya and labels me label_array and same in Test_data , test_labels

- esme Data_Matrix ka shape hai (n , d) means sare Data_points row_wise likhe hue hai par hame pehle ese (d,n) shape me convert karna hai means sare data_points ko column_wise likhna hai

- Question 1
    - How many examples are there in the training dataset?

In [113]:
# Examples bole to number of data_points to jab ham Training_data.shape karegenge to hamare paas aayega (n,d) then ham eska 0th index le lenge

Training_shape = Training_data.shape
print("Shape Of Training Data: ",Training_shape)
Num_Training_examples = Training_shape[0]
print("Total Examples: ",Num_Training_examples)

Shape Of Training Data:  (404, 13)
Total Examples:  404


- Question 2
    - How many examples are there in the test dataset?

In [114]:
Test_shape = Test_data.shape
print("Shape Of Test Data: ",Test_shape)

num_test_examples = Test_shape[0]
print("Total Examples: ",num_test_examples)

Shape Of Test Data:  (102, 13)
Total Examples:  102


- Question 3
    - How many features are there in the dataset?

In [115]:
Training_shape = Training_data.shape
print("Shape Of Training Data: ",Training_shape)

training_features = Training_shape[1]
print("Total Training Features: ",training_features)

Shape Of Training Data:  (404, 13)
Total Training Features:  13


### Data Modification

$$
h_w(x) = w_1 x_1 + w_2 x_2 + \dots + w_d x_d = x^T w
$$

### Definitions

- $x_i$: $i^{th}$ feature  
- $x$: Feature matrix of shape $(d, n)$  
- $w = [w_1, w_2, \dots, w_d]^T$: Weight vector  

---

The above model **always passes through the origin**.

To fix this, add an intercept term $w_0$ with a dummy feature $x_0 = 1$.

$$
h_w(x) = w_0 x_0 + w_1 x_1 + \dots + w_d x_d = x^T w
$$

---

- $x_0$ is the **dummy feature**
- Set $x_0 = 1$ for all examples

---

- $w$: shape $(d+1, 1)$  
- $x$: shape $(d+1, n)$  
- First row of $x$ → all 1s  

---

## Task

- Add the dummy feature to **Training_data** and **test_data**
- Use updated matrices for model training

**Note:** You may convert training data shape to $(d, n)$ if convenient.

In [116]:
# For Training_data
train_ones_array = np.ones([1 , Training_data.shape[0]])
# print(train_ones_array, "\n \n" ,len(train_ones_array))
print("Ones Array Shape: ",train_ones_array.shape)

print("Training Data Previous Shape: ",Training_data.shape)
Training_data = Training_data.T
print("Training Data Shape After Transpose: ",Training_data.shape)
x_train = np.vstack([train_ones_array , Training_data])
print("Training Data Shape After Stack Ones_array above it: ",x_train.shape)



Ones Array Shape:  (1, 404)
Training Data Previous Shape:  (404, 13)
Training Data Shape After Transpose:  (13, 404)
Training Data Shape After Stack Ones_array above it:  (14, 404)


In [117]:
# For Test_Data
test_ones_array = np.ones([1 , Test_data.shape[0]])
print("Ones Array Shape: " , test_ones_array.shape)

print("Shape of Test Data Before Transpose: ", Test_data.shape)
Test_data = Test_data.T
print("Shape of Test_Data after transpose: ",Test_data.shape)

x_test = np.vstack([test_ones_array , Test_data])
print("Shape Of test Data after adding Ones Row: ", x_test.shape)


Ones Array Shape:  (1, 102)
Shape of Test Data Before Transpose:  (102, 13)
Shape of Test_Data after transpose:  (13, 102)
Shape Of test Data after adding Ones Row:  (14, 102)


### Question 4

If the solution of optimization problem is obtained by setting the first derivative of loss function (squared loss) to zero, find the value of $w_0 + w_1 + \dots + w_d$.

---

$$
\hat{y} = w_0 + w_1 x_1 + w_2 x_2 + \dots + w_d x_d
$$

### Yahan:

- $x_1, x_2, \dots, x_d$ → features (crime rate, rooms, etc.)
- $w_0, w_1, \dots, w_d$ → weights jo hume find karne hain
- $w_0$ → bias term (intercept)

---

$$
L(w) = \sum_{i=1}^{n} (y_i - \hat{y}_i)^2
$$

$$
L(w) = (y - Xw)^T (y - Xw)
$$

---

$$
\frac{\partial L}{\partial w} = 0
$$

$$
-2 X^T (y - Xw) = 0
$$

$$
X^T y = X^T X w
$$

$$
w = (X^T X)^{-1} X^T y
$$

---

$$
w_0 + w_1 + \dots + w_d = \mathbf{1}^T w
$$

$$
= \mathbf{1}^T (X^T X)^{-1} X^T y
$$

In [123]:
X = x_train.T
Y = labels
print(X.shape , Y.shape)

XtX = X.T @ X
XtY = X.T @ Y

W = np.linalg.inv(XtX) @ XtY
print(W.shape)

Sum = np.sum(W)
print(Sum)



(404, 14) (404,)
(14,)
20.524582464673482


- Question 5
    - Find the average of the predictions made by the above model.

In [127]:
print(W.shape , x_train.shape)

prediction = W.T @ x_train
print(prediction.shape)
prediction_avg = np.mean(prediction)
print(prediction_avg)

(14,) (14, 404)
(404,)
22.309158415843903


### Question 6

Find the loss for the training data points using the above model.  
Consider the loss to be defined as:

$$
\sqrt{\frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2}
$$

Where $\hat{y}_i$ is the prediction for the $i^{th}$ data point.

In [131]:
prediction = W.T @ x_train
error = labels - prediction
error_sq = error**2
MSE = np.mean(error_sq)
Rote_MSE = np.sqrt(MSE)
print(Rote_MSE)

4.552387969840813


### Question 7

Find the loss for the test data points using the above model.  
Consider the loss to be defined as:

$$
\sqrt{\frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2}
$$

Where $\hat{y}_i$ is the prediction for the $i^{th}$ data point.

In [132]:
prediction_test = W.T @ x_test
error = test_labels - prediction_test
error_sq = error ** 2
MSE = np.mean(error_sq)
Rote_MSE = np.sqrt(MSE)
print(Rote_MSE)

5.327662216181741


### Question 8

Find the weights using gradient descent.  
Use a constant learning rate of $\eta = 10^{-10}$.

- Initialize the weight vector as a zero vector  
- Update the weights for 100 iterations  

Enter the **sum of all the weights**.

### Solution

### Gradient Descent Formula

$$
W \leftarrow W - 2\eta \left[(XX^T)W - XY\right]
$$

---


### Step by step solution:

**Step 1 — Initialize:**

$$
W =
\begin{bmatrix}
0 \\
0 \\
\vdots \\
0
\end{bmatrix}
\quad \text{shape: (14,)}
$$

---

**Step 2 — Gradient calculate karo:**

$$
\text{gradient} = (XX^T)W - XY
$$

- $(XX^T)W \rightarrow (14,14) \times (14,) = (14,)$  
- $XY \rightarrow (14,404) \times (404,) = (14,)$  

Final gradient shape: $(14,)$

---

**Step 3 — Update karo:**

$$
W \leftarrow W - 2\eta \cdot \text{gradient}
$$

---

**Step 4 — 100 baar repeat karo**

---

**Step 5 — Answer:**

$$
\text{ans} = \sum_{i=0}^{d} w_i = w_0 + w_1 + \dots + w_{13}
$$

In [149]:
X = x_train
print(X.shape)
Y = labels
print(Y.shape)
eta = 1e-10
W = np.zeros(X.shape[0])
print(W.shape)

for i in range(100):
    gradient = ( X @ X.T ) @ W - X@Y
    W = W - 2*eta*(gradient)
Sum = np.sum(W)
print(Sum)


(14, 404)
(404,)
(14,)
0.0589590611959026


### Question 9

Find the loss for the training data points if the model is learnt using gradient descent (as in Question 8).  
Consider the loss to be defined as:

$$
\sqrt{\frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2}
$$

Where $\hat{y}_i$ is the prediction for the $i^{th}$ data point.

In [155]:
print(W.shape , x_train.shape)

prediction = W.T @ x_train

error = labels - prediction
error_sq = error ** 2
MSE = np.mean(error_sq)
Rote_MSE = np.sqrt(MSE)
print(Rote_MSE)

(14,) (14, 404)
11.13727323702196


### Question 10

Find the loss for the test data points if the model is learnt using gradient descent (as in Question 8).  
Consider the loss to be defined as:

$$
\sqrt{\frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2}
$$

Where $\hat{y}_i$ is the prediction for the $i^{th}$ data point.

In [156]:
prediction = W.T @ x_test
error = test_labels - prediction
error_sq = error ** 2
MSE = np.mean(error_sq)
Rote_MSE = np.sqrt(MSE)
print(Rote_MSE)

10.964491250062148
